tmt/base.py

```diff
@@ -375,10 +377,101 @@ def from_spec(cls, raw: _RawRequireFmfId) -> 'RequireFmfId':  # type: ignore[ove
+    type: Optional[str]
+    pattern: Optional[List[str]]
+
+
+@dataclasses.dataclass
+class RequireFile(
+        tmt.utils.SpecBasedContainer,
+        tmt.utils.SerializableContainer,
+        tmt.export.Exportable['RequireFile']):
+    VALID_KEYS: ClassVar[List[str]] = ['type', 'pattern']
+
+    type: Optional[str] = None
+    pattern: Optional[List[str]] = None
+
+    # ignore[override]: expected, we do want to return more specific
+    # type than the one declared in superclass.
+    def to_dict(self) -> _RawRequireFile:  # type: ignore[override]
+        """ Return keys and values in the form of a dictionary """
+        return cast(_RawRequireFile, super().to_dict())
+
+    # ignore[override]: expected, we do want to return more specific
+    # type than the one declared in superclass.
+    def to_minimal_dict(self) -> _RawRequireFile:  # type: ignore[override]
+        """ Convert to a mapping with unset keys omitted """
+        return cast(_RawRequireFile, super().to_minimal_dict())
+
+    # ignore[override]: expected, we do want to return more specific
+    # type than the one declared in superclass.
+    def to_spec(self) -> _RawRequireFile:  # type: ignore[override]
+        """ Convert to a form suitable for saving in a specification file """
+        return self.to_dict()
+
+    # ignore[override]: expected, we do want to return more specific
+    # type than the one declared in superclass.
+    def to_minimal_spec(self) -> _RawRequireFile:  # type: ignore[override]
+        """ Convert to specification, skip default values """
+        return cast(_RawRequireFile, super().to_minimal_spec())
+
+    # ignore[override]: expected, we do want to accept and return more
+    # specific types than those declared in superclass.
+    @classmethod
+    def from_spec(cls, raw: _RawRequireFile) -> 'RequireFile':
+        """ Convert from a specification file or from a CLI option """
+        return RequireFile(
+            **{key: cast(Optional[str], raw.get(key, None)) for key in cls.VALID_KEYS})
+
+    def validate(self) -> Tuple[bool, str]:
+        """
+        Validate file requirement and return a human readable error
+
+        Return a tuple (boolean, message) as the result of validation.
+        The boolean specifies the validation result and the message
+        the validation error. In case the file requirement is valid, return an empty
+        string as the message.
+        """
+        if self.type and self.type not in ['library', 'file']:
+            return False, 'Type must be library or file'
+        if self.pattern:
+            if not isinstance(self.pattern, str) and not isinstance(self.pattern, list):
+                return False, 'Pattern must be single string or list'
+            if isinstance(self.pattern, list):
+                for path in self.pattern:
+                    if not isinstance(path, str):
+                        return False, f'Path {path} must be string'
+        return True, ''
```

Comment:
Note for me: too late. We need to find a way to employ normalization here, to benefit from its powers.